### PR TITLE
chore(ci): test TestModels using parameterized JUnit tests

### PR DESCRIPTION
**Pull request by robin-aws**
*Description of changes:*
Replaces the GitHub actions matrices with `@ParameterizedTest`s that run the same commands for every discovered test model, unless explicitly marked as disabled. 

There are several advantages to this:
* Makes it possible to run all tests locally much more easily, using standard JUnit tooling such as IntelliJ test support.
* Makes it more feasible to maintain branch protection that requires all checks, since the names won't change as test models are added or enabled for new targets.
* Ensures every test model is verified and tested against each supported language by default, unless marked as not supported yet, in which case it will be SKIPPED.
* More efficiently uses GHA runners by not requiring a separate runner for every individual test: the tests are sharded across a tunable number of runners. This also makes it more feasible to test on multiple OS' in the future, which will also increase the number of jobs.
* Opens the door for utilities like "update the patch files for every test model currently running against Rust".

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .github/workflows/format.yml
Comment (texastony) [inline .github/workflows/format.yml:1]: Why are we modifying this?

File: .github/workflows/pull.yml
Comment (texastony) [inline .github/workflows/pull.yml:1]: Why are we disabling Rust CI?

File: .github/workflows/test_models_dafny_verification.yml [hunk 1]
```diff
@@ -95,19 +56,11 @@ jobs:
...
-        working-directory: ./${{ matrix.library }}
-        run: |
-          make dafny-reportgenerator
+      - name: Execute smithy-dafny-codegen-test tests
+        uses: gradle/gradle-build-action@v2
+        env:
+          JUNIT_SHARD: ${{ matrix.shard }}
+          JUNIT_SHARD_COUNT: ${{ inputs.num_shards }}
+        with:
+          arguments: :smithy-dafny-codegen-test:test --tests '*smithydafny*' --info
```
Comment (texastony) [inline .github/workflows/test_models_dafny_verification.yml:65]: Does this gradle build action trigger Dafny verification?

Or is it just running tests?

The command is not clear.

File: .github/workflows/test_models_net_tests.yml [hunk 2]
```diff
@@ -106,27 +76,11 @@ jobs:
...
-        shell: bash
-        run: |
-          make test_net
+      - name: Execute smithy-dafny-codegen-test tests
+        uses: gradle/gradle-build-action@v2
+        env:
+          JUNIT_SHARD: ${{ matrix.shard }}
+          JUNIT_SHARD_COUNT: ${{ inputs.num_shards }}
+        with:
+          arguments: :smithy-dafny-codegen-test:test --tests '*smithydotnet*' --info
```
Comment (texastony) [inline .github/workflows/test_models_net_tests.yml:85]: So... Gradle is shelling out to `dotnet`?

File: codegen/smithy-dafny-codegen-test/bin/test/software/amazon/polymorph/TestModelTest.class
Comment (texastony) [inline codegen/smithy-dafny-codegen-test/bin/test/software/amazon/polymorph/TestModelTest.class:1]: Why are we committing the compiled Java?

File: codegen/smithy-dafny-codegen-test/bin/test/software/amazon/polymorph/smithydafny/DafnyTestModels.class
Comment (texastony) [inline codegen/smithy-dafny-codegen-test/bin/test/software/amazon/polymorph/smithydafny/DafnyTestModels.class:1]: Why are we committing the compiled Java?

File: codegen/smithy-dafny-codegen-test/bin/test/software/amazon/polymorph/smithydotnet/DotnetTestModels.class
Comment (texastony) [inline codegen/smithy-dafny-codegen-test/bin/test/software/amazon/polymorph/smithydotnet/DotnetTestModels.class:1]: Why are we committing the compiled Java?

File: codegen/smithy-dafny-codegen-test/bin/test/software/amazon/polymorph/smithyjava/JavaTestModels.class
Comment (texastony) [inline codegen/smithy-dafny-codegen-test/bin/test/software/amazon/polymorph/smithyjava/JavaTestModels.class:1]: Why are we committing the compiled Java?

File: codegen/smithy-dafny-codegen-test/src/test/java/software/amazon/polymorph/smithydafny/DafnyTestModels.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        DafnyVersion dafnyVersion = DafnyVersion.parse(System.getenv("DAFNY_VERSION"));
+        if (dafnyVersion.compareTo(DafnyVersion.parse("4.4.0")) < 0) {
+            Assumptions.assumeFalse(TESTS_THAT_NEED_DAFNY_4_4.contains(relativeTestModelPath));
+        }
+
+        Path testModelPath = getTestModelPath(relativeTestModelPath);
+        make(testModelPath, "polymorph_dafny");
+        make(testModelPath, "verify", "CORES=" + Runtime.getRuntime().availableProcessors());
+        make(testModelPath, "dafny-reportgenerator");
+    }
```
Comment (texastony) [inline codegen/smithy-dafny-codegen-test/src/test/java/software/amazon/polymorph/smithydafny/DafnyTestModels.java]: OK. Yeah. Gradle is shelling out to Dafny to Verify.
Groovy.

File: codegen/smithy-dafny-codegen-test/src/test/java/software/amazon/polymorph/smithydotnet/DotnetTestModels.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    @ParameterizedTest
+    @MethodSource("discoverTestModels")
+    void testModelsForDotnet(String relativeTestModelPath) {
+        Assumptions.assumeFalse(DISABLED_TESTS.contains(relativeTestModelPath));
+
+        Path testModelPath = getTestModelPath(relativeTestModelPath);
+        make(testModelPath, "polymorph_dafny");
+        make(testModelPath, "polymorph_dotnet");
+        make(testModelPath, "transpile_net");
+        make(testModelPath, "test_net");
```
Comment (texastony) [inline codegen/smithy-dafny-codegen-test/src/test/java/software/amazon/polymorph/smithydotnet/DotnetTestModels.java]: And then shelling out to .NET.

**Review by texastony — COMMENTED** [review]
I've wrapped my head around everything except:
- Why commit compiled java classes?
- Why disable formatting checks?
- Why disable Rust CI?

**Comment by robin-aws** [comment]
Those were all unintentional/temporary changes I neglected to undo. Fixed now (especially by applying the same changes for Rust). 

Apologies for the sloppy PR, I don't do my usual review of changes before flipping out of draft.

**Review by texastony — APPROVED** [review]
Nice!

**Pull request merged**